osquery/tables/smart/smart_drives.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <smartmontools/libsmartctl.h>
+#include <smartmontools/smartctl_errs.h>
+
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+#include <osquery/tables.h>
+
+#include "osquery/events/linux/udev.h"
+
+namespace osquery {
+namespace tables {
+
+/* @brief Represents a hardware driver type that SMART api can you use to query
+ * device information.
+ *
+ * @param driver name of SMART controller driver
+ * @param maxID max ID number of which disks on the controller is monitored
+ */
+struct hardwareDriver {
+  std::string driver;
+  size_t maxID;
+};
+
+std::ostream& operator<<(std::ostream& stream,
+                         const std::vector<hardwareDriver>& devs) {
+  for (const auto& dev : devs) {
+    stream << " " << dev.driver;
+  }
+
+  return stream;
+}
+
+/// Static map of supported controller types.
+static const std::map<std::string, hardwareDriver>
+    kSMARTExplicitDriverToDevice = {
+        {"megaraid_sas", hardwareDriver{"megaraid,", 127}},
+        {"hpsa", hardwareDriver{"cciss,", 14}},
+};
+
+/// Gets all devices of class 'Mass storage controller'.
+std::vector<std::string> getStorageCtlerClassDrivers() {
+  std::vector<std::string> results;
+
+  auto devices = SQL::selectAllFrom("pci_devices");
+  for (const auto& device : devices) {
+    if (device.find("pci_class") != device.end() &&
+        device.at("pci_class") == "Mass storage controller" &&
+        device.find("driver") != device.end()) {
+      auto i =
```

Comment:
super nit: seems like `i` might be a poor choice for variable name here, and should instead be reserved for iterations >.>